src/gui/qgscolorwidgets.cpp

```diff
@@ -1153,28 +1157,28 @@ void QgsColorRampWidget::paintEvent( QPaintEvent *event )
     //draw background ramp
     for ( int c = 0; c <= maxValue; ++c )
     {
-      int colorVal = static_cast<int>( componentRange() * static_cast<double>( c ) / maxValue );
+      float colorVal = static_cast<float>( c ) / static_cast<float>( maxValue );
       //vertical sliders are reversed
       if ( mOrientation == QgsColorRampWidget::Vertical )
       {
-        colorVal = componentRange() - colorVal;
+        colorVal = 1.f - colorVal;
       }
-      alterColor( color, mComponent, colorVal );
-      if ( color.hue() < 0 )
+      alterColorF( color, mComponent, colorVal );
+      if ( color.hueF() < 0 )
       {
-        color.setHsv( hue(), color.saturation(), color.value() );
+        color.setHsvF( hueF(), color.saturationF(), color.valueF() );
       }
       pen.setColor( color );
       painter.setPen( pen );
       if ( mOrientation == QgsColorRampWidget::Horizontal )
       {
         //horizontal
-        painter.drawLine( QLineF( c + mMargin, mMargin, c + mMargin, height() - mMargin - 1 ) );
+        painter.drawLine( c + mMargin, mMargin, c + mMargin, height() - mMargin - 1 );
```

Comment:
We still need the QLineF here, as the drawLine methods which use x/y directly truncate to int and will cause hi-dpi issues:

```suggestion
        painter.drawLine( QLineF( c + mMargin, mMargin, c + mMargin, height() - mMargin - 1 ) );
```

Reply:
after some clang-tidy warnings, I reworked this part to use only integers. There is no use of float here because we draw a colored line for each pixel of the ramp regarding its width/height.

Reply:
That can cause hipdi issues though -- the qpainter methods which use int have different internal logic vs the float types

Reply:
`drawLine(int,int,int,int)` calls eventually `drawLines(QLine*,int)` which converts the integer to float before calling `drawLines(QLineF*, int)` (see [here](https://code.qt.io/cgit/qt/qtbase.git/tree/src/gui/painting/qpaintengine.cpp?h=71bc951870906abbe656d9c2f08c42c3bf35c554#n800) ).

So it looks to me that the internal logic is the same actually, and there is no need to make the conversion beforehand.

Reply:
But that's assuming that the platform is using that base class QPaintEngine method without overriding it, and I'm not confident that's always the case.

The qt 5 docs state (from https://doc.qt.io/qt-5/highdpi.html , under "Migrate Existing Applications"):

"In the long term, the application should be adapted to run unmodified:

Always use the qreal versions of the [QPainter](https://doc.qt.io/qt-5/qpainter.html) drawing API."

This note has been removed from the qt 6 docs (https://doc.qt.io/qt-6/highdpi.html), so maybe this isn't necessary in qt 6 builds. But for now we should follow the qt 5 recommendations.

This isn't just theoretical by the way -- the QLineF here was introduced some time ago to fix a real hidpi issue. See https://github.com/qgis/QGIS/commit/67f5022c083a57bbdfa7e5eac03d87c8fe95e185, https://github.com/qgis/QGIS/issues/23900

Reply:
Done [33438f1](https://github.com/qgis/QGIS/pull/58375/commits/33438f1c61f4badf73552807059a435fc99f3669)